### ISSUE TITLE
Restructure docs

### ISSUE THREAD

**Issue by ddeboer** [zotonic/zotonic#1133]
I’d like to restructure the docs so it will be easier for people to find what they’re looking for. Before I move in with a a PR, however, I’d like discuss my proposal with you.

Main changes:
* Add a ‘user guide’ and call the other part ‘developer guide’. An introduction to Zotonic in layman’s terms (perhaps with some more screenshots) will appeal to both end users and developers that prefer an easy-to-read intro.
* Add a ‘best practices’ section, as there’s quite some demand for core team recommendations about how to solve problems.
* Remove chapters that are either empty or very low on informative value.

---

1. User guide
  - Why Zotonic?
  - Zotonic CMS
  - Data model
    - Resources & categories
    - Edges & predicates
    - Media files
    - Users & identities
  - Submitting issues and feature requests

2. Developer guide

  - Introduction: CMS and framework
  - Installation and configuration
  - Sites
    - Structure
    - Configuration
  - Modules
  - Controllers
  - Dispatcher
  - Data
    - Custom properties (pivot)
  - Templates
    - Template lookup
    - Inheritance and overriding
    - Tags
    - Filters
    - Forms and validation [**edited**]
    - Models
      - Search
    - Media files
    - Actions [**edited**]
    - Adding CSS and JavaScript [**edited**]
    - JavaScript
  - Security
    - Authentication
    - Authorization
  - Translation
  - Notifications
    - Observing
    - Sending
  - Wires (transporting data between browser and server)
  - Publish/subscribe (MQTT)
  - API services
  - E-mail
  - Contributing
    - Pull requests
    - Testing
    - Coding style 
    - Contributing documentation

3. Cookbooks
  - Frontend templates
  - The admin
  - Command line
  
4. Best practices
  - Template best practices
  - ACL best practices
  - Notification best practices

5. Reference 
  - Configuration
  - Built-in notifications
  - etc.

**Comment by mworrell** [comment]
The structure looks logical.
Maybe after this we can insert a Quick Start before the Developer Guide.

Pulling in some others: @AlainODea @kaos @michielklonhammer

**Comment by ddeboer** [comment]
@mworrell I’m all for a Quick Start or Getting Started chapter; it should be the first one in the Developer Guide. It’s probably easiest to distill that chapter after we’ve cleaned up the current docs.

**Comment by ArthurClemens** [comment]
I like this structure. 

I would add to Templates:

* Actions
* Adding CSS and Javascript
* Forms and validation

I would add to Reference:

* Internationalization
* Users, Authentication and Access control

A separate section **Writing your own module**:

* Module structure
* Listening for events
* Adding an admin menu item
* Auto creation of structure / manage_schema => update/extend "Module versioning"
* Creating your own API

**Comment by AlainODea** [comment]
I agree with these changes.  Separate User and Developer Guides fit with the conventional organization of software framework docs and I think it will improve traction and success for CMS end users vs app developers.

I would certainly incorporate @ArthurClemens recommendations, but **Writing  your own module** would live toward the end of the Developet Guide.

Audience definitions are *extremely* important here.  We need to clearly define what we mean by User and Developer for these guides and provide a set of criteria or questions that readers can quicklyuse to know if they are in the right place.  Developer Guide may be confusing as some may think it's a Contributor Guide.

I'm happy to review as this goes along.

**Comment by ddeboer** [comment]
> A separate section Writing your own module

Definitely something that we should add to the docs! Perhaps have it a as a cookbook entry?

**Comment by kaos** [comment]
Great initiative!

**Comment by ddeboer** [comment]
What about [Driebit’s Ginger guide](http://support.ginger.nl/en/page/469/structuur) as a starting point for the user guide?

**Comment by mworrell** [comment]
Looks like a good start - we shouldn't reinvent wheels (and chapter structures)

**Comment by mmzeeman** [comment]
That is really nice.

**Comment by mworrell** [comment]
Maybe we can also refresh the css and navigation menu of the docs.
I find it a bit hard to go up and down in the navigation (and yes, I know it is on top... I am always looking at the left hand menu to go up...)

**Comment by mworrell** [comment]
I was also thinking of building a real full-text index, mirroring the documentation pages onto Zotonic internal pages. Then we can offer a single search function of the site (blog articles etc) and the generated docs.  The pages could be inserted with the correct page_path, making it easy to generate this "mirror" and also easy to keep it up to date.

**Comment by ddeboer** [comment]
Why not use https://readthedocs.org or something similar? Let’s spend our time on the contents rather than the CSS and hosting of the docs.

**Comment by mmzeeman** [comment]
A little bit of "Chew your own dog food" might be beneficial for other things.

**Comment by mworrell** [comment]
What I propose with indexing is a generic solution for indexing a static site, so also useful to add some dog food.

**Comment by mworrell** [comment]
Maybe we can use the Sphinx theme (with some minor Zotonic styling adaptations), it is quite nice. How is the license of that one?

**Comment by mworrell** [comment]
Found it https://github.com/snide/sphinx_rtd_theme
It MIT licensed - so we can use it.

**Comment by ddeboer** [comment]
@ArthurClemens 

> I would add to Templates:

> Actions 
> Adding CSS and Javascript

Added (I edited my first post above).

> Forms and validation

That’s what I called ‘Validators’, right? I do prefer ‘Forms and validation’ though.

> I would add to Reference:

> Internationalisation
> Users, Authentication and Access control

These are more thematic groups of references, right? What kind of reference would you expect here? I think Internationalisation comes down to template tags mostly, which is a separate reference chapter.

**Comment by ArthurClemens** [comment]
Yes, groups of references. I think it is about grouping information so that it becomes clear what the possibilities are, which tasks will be for whom, pitfalls and such.

For instance on Internationalisation:

* Enabling languages in the admin
* URLs / dispatch with or without language code
* Templates: multi-language tags
* When creating custom templates or modules: generate POT files, update (this page exists)
* Localisable paths (which is WIP or concept I believe)

**Comment by ddeboer** [comment]
Under a ‘Zotonic Reference’ part I understand lists of all configuration parameters, template tags, filters  etc. that are available in Zotonic.

The first part of the Developer Guide (perhaps Guides) is meant as more thematic introductions to the various topics. Your Internationalisation example better fits that first part; it would make a great Translation chapter. 

So for instance the Developer Guide’s Template chapter will not contain a list of all template tags; those will go into a Reference entry. Instead, the Template chapter will be an introduction and overview of all things template, including how to use tags and some tag examples. Does that make sense?

**Comment by ArthurClemens** [comment]
You are right, better to make a clear distinction.

**Comment by ddeboer** [comment]
For all those that want to help out reviewing the changes: have a look at #1149.
